crates/ra_hir/src/ty/lower.rs

```diff
@@ -199,13 +201,28 @@ pub(super) fn substs_from_path_segment(
         }
     }
     // add placeholders for args that were not provided
-    // FIXME: handle defaults
     let supplied_params = substs.len();
     for _ in supplied_params..def_generics.count_params_including_parent() {
         substs.push(Ty::Unknown);
     }
     assert_eq!(substs.len(), def_generics.count_params_including_parent());
-    Substs(substs.into())
+    let mut substs = Substs(substs.into());
+
+    // handle defaults
+    if let Some(def_generic) = def_generic {
+        let default_substs = db.generic_defaults(def_generic);
+        assert_eq!(substs.len(), default_substs.len());
+
+        let mut i = 0;
+        substs.walk_mut(&mut |ty| {
```

Comment:
walk_mut does a deep walk, you don't want that here. How about instead not turning the `Vec` into a `Substs` yet by leaving out line 209, then you can just do a normal for loop here?

Reply:
Sure and Done :)